vim-patch:8.1.{1734,1742,1743}: highlight, match and screen.c refactoring

lewis6991 opened the pull request:
```
    vim-patch:8.1.1734: the evalfunc.c file is too big

    Problem:    The evalfunc.c file is too big.
    Solution:   Move some functions to other files.
```
https://github.com/vim/vim/commit/29b7d7a9aac591f920edb89241c8cde27378e50b

```
    vim-patch:8.1.1742: still some match functions in evalfunc.c

    Problem:    Still some match functions in evalfunc.c.
    Solution:   Move them to highlight.c.
```

https://github.com/vim/vim/commit/7dfb016d25e3e3e1f4411026dda21d1536f21acc

```
    vim-patch:8.1.1743: 'hlsearch' and match highlighting in the wrong place

    Problem:  'hlsearch' and match highlighting in the wrong place.
    Solution: Move highlighting from inside screen functions to highlight.c.
```
https://github.com/vim/vim/commit/bbca7732e8a3deb6e5dcf84739579a2667a75475

lewis6991 committed:
vim-patch:8.1.1734: the evalfunc.c file is too big

Problem:    The evalfunc.c file is too big.
Solution:   Move some functions to other files.

https://github.com/vim/vim/commit/29b7d7a9aac591f920edb89241c8cde27378e50b

seandewar reviewed (approved):
LGTM I think. Didn't see any (relevant) sneaky undocumented changes made in the patch.
Might also be worth porting 8.1.1742 at some point if it's possible

lewis6991 committed:
vim-patch:8.1.1743: 'hlsearch' and match highlighting in the wrong place

Problem:  'hlsearch' and match highlighting in the wrong place.
Solution: Move highlighting from inside screen functions to highlight.c.

https://github.com/vim/vim/commit/bbca7732e8a3deb6e5dcf84739579a2667a75475

zeertzjq commented on src/nvim/screen.c:
```diff
@@ -2599,66 +2591,12 @@ static int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool noc
+    area_highlighting |= prepare_search_hl_line(wp, lnum, (colnr_T)v,
+                                                &line, &search_hl, &search_attr,
+                                                &search_attr_from_match, has_fold, end_fill);
+    ptr = line + v; // "line" may have been updated
```
```suggestion
    ptr = line + v;  // "line" may have been updated
```

lewis6991 committed:
vim-patch:8.1.1742: still some match functions in evalfunc.c

Problem:    Still some match functions in evalfunc.c.
Solution:   Move them to highlight.c.

https://github.com/vim/vim/commit/7dfb016d25e3e3e1f4411026dda21d1536f21acc

lewis6991 commented:
> Might also be worth porting 8.1.1742 at some point if it's possible

Done

zeertzjq commented:
a lot of crashes in CI

lewis6991 commented:
Yep, debugging.

lewis6991 commented:
Was missing a couple of pointer derefs. Should be good now.

zeertzjq commented:
`match_add()` and `match_delete()` can be made static now.

lewis6991 commented:
I'm going to be moving those functions again with 8.2.1078 after this so I don't think it's worth it.

zeertzjq commented on src/nvim/highlight_group.c:
```diff
@@ -3077,6 +3081,508 @@ matchitem_T *get_match(win_T *wp, int id)
+                           match_T *search_hl, int *search_attr, bool *search_attr_from_match,
+                           bool has_fold, bool end_fill)
+{
+  // Handle highlighting the last used search pattern and matches.
+  // Do this for both search_hl and the match list.
```
I feel this comment looks a bit strange. Maybe it is better to put this above the `while`.

zeertzjq commented on src/nvim/screen.c:
```diff
@@ -2599,66 +2591,12 @@ static int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool noc
-    }
+  if (!number_only) {
+    v = (long)(ptr - line);
+    area_highlighting |= prepare_search_hl_line(wp, lnum, (colnr_T)v,
+                                                &line, &search_hl, &search_attr,
+                                                &search_attr_from_match, has_fold, end_fill);
```
Maybe it is better to skip this function call if either of `has_fold` or `end_fill` is true.

seandewar commented on src/nvim/highlight_group.c:
```diff
@@ -2797,3 +2807,1164 @@ int name_to_ctermcolor(const char *name)
+{
+  matchitem_T *cur = wp->w_match_head;  // points to the match list
+  match_T     *shl;                     // points to search_hl or a match
+  int          shl_flag = false;        // flag to indicate whether search_hl
```
bool

lewis6991 commented:
Thanks for the reviews. Think I got all the comments.

zeertzjq commented:
I think it is better to have a rebase now.
